### PR TITLE
Implementation of Deserialize{Context,Register}

### PR DESCRIPTION
**Pull request by Shimuuar**
It's still partial implementation. There's no evaluation and types could be wrong. 

I'm still confused how it should be evaluated and wasn't able to find corresponding code in scala version

### CODE REVIEW

File: ergotree-ir/src/mir/deserialize_context.rs [hunk 1]
```diff
@@ -0,0 +1,82 @@
+//! Extracts context variable as byte array
+
+use crate::serialization::op_code::OpCode;
+use crate::serialization::sigma_byte_reader::SigmaByteRead;
+use crate::serialization::sigma_byte_writer::SigmaByteWrite;
+use crate::serialization::SerializationError;
+use crate::serialization::SigmaSerializable;
+use crate::types::stype::SType;
+
+/// Extracts context variable as byte array
```
Comment (greenhat) [inline ergotree-ir/src/mir/deserialize_context.rs:10]:
```suggestion
/// Extracts context variable as Coll[Byte], deserializes it to script and then executes this script in the current context.
/// The original `Coll[Byte]` of the script is available as `getVar[Coll[Byte]](id)`
/// On evaluation returns the result of the script execution in the current context
```

File: ergotree-ir/src/mir/deserialize_context.rs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub struct DeserializeContext {
+    /// Type of variable
```
Comment (greenhat) [inline ergotree-ir/src/mir/deserialize_context.rs:13]:
```suggestion
    /// Result type of the deserialized script
```

File: ergotree-ir/src/mir/deserialize_context.rs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    pub ctx_tpe: SType,
```
Comment (greenhat) [inline ergotree-ir/src/mir/deserialize_context.rs:14]:
```suggestion
    pub tpe: SType,
```

File: ergotree-ir/src/mir/deserialize_context.rs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    /// Context ID
```
Comment (greenhat) [inline ergotree-ir/src/mir/deserialize_context.rs:15]:
```suggestion
    /// identifier of the context variable
```

File: ergotree-ir/src/mir/deserialize_context.rs [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    pub ctx_id: u8,
```
Comment (greenhat) [inline ergotree-ir/src/mir/deserialize_context.rs:16]:
```suggestion
    pub id: u8,
```

File: ergotree-ir/src/mir/deserialize_context.rs [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+}
+impl DeserializeContext {
+    pub(crate) const OP_CODE: OpCode = OpCode::DESERIALIZE_CONTEXT;
+
+    /// FIXME: Is this true??
```
Comment (greenhat) [inline ergotree-ir/src/mir/deserialize_context.rs:21]: Yes, it's the result type of the deserialized script execution.

File: ergotree-ir/src/mir/deserialize_register.rs [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+use crate::serialization::sigma_byte_writer::SigmaByteWrite;
+use crate::serialization::SerializationError;
+use crate::serialization::SigmaSerializable;
+use crate::types::stype::SType;
+
+/// Extract register of SELF box as byte array, deserialize it into Value and inline into executing script.
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub struct DeserializeRegister {
+    /// Register number
+    pub reg_n: u8,
```
Comment (greenhat) [inline ergotree-ir/src/mir/deserialize_register.rs:15]:
```suggestion
    pub reg_id: u8,
```

File: ergotree-ir/src/mir/deserialize_register.rs [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    /// Type of value in register
+    pub reg_tpe: SType,
```
Comment (greenhat) [inline ergotree-ir/src/mir/deserialize_register.rs:17]:
```suggestion
    pub tpe: SType,
```

File: ergotree-ir/src/mir/deserialize_register.rs [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    /// FIXME:
+    pub reg_val: Option<Box<Expr>>,
```
Comment (greenhat) [inline ergotree-ir/src/mir/deserialize_register.rs:19]:
```suggestion
    pub default: Option<Box<Expr>>,
```

File: ergotree-ir/src/mir/deserialize_register.rs [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+}
+impl DeserializeRegister {
+    pub(crate) const OP_CODE: OpCode = OpCode::DESERIALIZE_REGISTER;
+
+    /// FIXME: Is this true??
```
Comment (greenhat) [inline ergotree-ir/src/mir/deserialize_register.rs:24]: Yes. The type of value stored in the register.

**Comment by greenhat** [comment]
Thanks!

 > I'm still confused how it should be evaluated and wasn't able to find corresponding code in scala version

Yes, these IR nodes are more complicated.

They both are handled before the script execution. See - https://github.com/ScorexFoundation/sigmastate-interpreter/blob/6c51c13f7a494a191a7ea5645e56b04fb46a418d/sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala#L85-L102
and
https://github.com/ScorexFoundation/sigmastate-interpreter/blob/6c51c13f7a494a191a7ea5645e56b04fb46a418d/sigmastate/src/main/scala/org/ergoplatform/ErgoLikeInterpreter.scala#L21-L40
 
I think we can skip their evaluation for now. I have not thought about it yet, feel free to take a stab at it if you want though.

Regarding the names. We're trying to have consistent naming between Scala and Rust versions. So, if code translates "directly" its better to use Scala names.

**Comment by Shimuuar** [comment]
I updated field names to mirror scala better. 

> I think we can skip their evaluation for now. I have not thought about it yet, feel free to take a stab at it if you want though.

Let skip evaluation then. After all main main goal at the moment to be able to parse whole ergo blockchain

**Review by greenhat — APPROVED** [review]
Great! There is one commit suggestion (comment change) left unaccepted.

**Comment by Shimuuar** [comment]
Yep. Missed that one. Updated

**Comment by greenhat** [comment]
@Shimuuar Resolve the merge conflict(after merged GetVar), please.

**Comment by Shimuuar** [comment]
Done.

**Pull request merged**